src/nytid/cli/hr.nw

```diff
@@ -102,79 +102,100 @@ import appdirs
+    logging.warning(f"Can't load LADOK credentials: {err}")
+    _ladok_session = None
+  
+  return _ladok_session
 @
 
 There are two errors that can occur here.
 The first is that we can't import the [[ladok3]] module.
 The second is that we can't load the credentials.
 In either case, we can't create the [[ladok_session]] object.
 <<handle ladok setup errors>>=
-except ImportError as err:
-  logging.warning(f"Can't import ladok3, not using LADOK data: {err}")
-  ladok_session = None
-except Exception as err:
-  logging.warning(f"Can't load LADOK credentials: {err}")
-  ladok_session = None
 @
 
 \subsection{Setting up Canvas}
 
 We also need to query user information from Canvas to find the LADOK ID.
 So we should also set up Canvas.
-Similarly to LADOK, we'll set up a [[canvas_session]] object.
+We'll create a function to lazily set up a [[canvas_session]] object.
+This way, we only try to load credentials when they're actually needed.
 <<import and set up Canvas>>=
-try:
-  import canvasapi
-  import canvaslms.cli
+# Global variable to cache the canvas session
+_canvas_session = None
+
+def get_canvas_session():
+  """Lazily initialize and return the Canvas session."""
+  global _canvas_session
+  if _canvas_session is not None:
+    return _canvas_session
+  
+  try:
+    import canvasapi
+    import canvaslms.cli
 
-  dirs_canvas = appdirs.AppDirs("canvaslms",
-                                "dbosk@kth.se")                                   
+    dirs_canvas = appdirs.AppDirs("canvaslms",
+                                  "dbosk@kth.se")                                   
 
-  canvaslms_config = canvaslms.cli.read_configuration(
-                        f"{dirs_canvas.user_config_dir}/config.json")
+    canvaslms_config = canvaslms.cli.read_configuration(
+                          f"{dirs_canvas.user_config_dir}/config.json")
 
-  CANVAS_SERVER, CANVAS_TOKEN = canvaslms.cli.login.load_credentials(
-                                    canvaslms_config)
+    CANVAS_SERVER, CANVAS_TOKEN = canvaslms.cli.login.load_credentials(
+                                      canvaslms_config)
 
-  if CANVAS_SERVER and CANVAS_TOKEN:
-    canvas_session = canvasapi.Canvas(os.environ["CANVAS_SERVER"], 
-                                      os.environ["CANVAS_TOKEN"])
-  else:
-    canvas_session = None
-    logging.warning("Can't load Canvas credentials, run `canvaslms login`")
-<<handle canvas setup errors>>
+    if CANVAS_SERVER and CANVAS_TOKEN:
+      _canvas_session = canvasapi.Canvas(os.environ["CANVAS_SERVER"], 
+                                        os.environ["CANVAS_TOKEN"])
+    else:
+      _canvas_session = None
+      logging.warning("Can't load Canvas credentials, run `canvaslms login`")
+  except ImportError as err:
+    logging.warning(f"Can't import Canvas: {err}")
+    _canvas_session = None
+  except Exception as err:
+    logging.warning(f"Can't load Canvas credentials: {err}")
+    _canvas_session = None
```

Comment:
Same here. It's better to keep this in the `<<handle canvas setup errors>>` code block.